File: 0992_Subarrays with K Different Integers/SubarrayswithKDifferentIntegers_2.cs
public class Solution {
    public int SubarraysWithKDistinct(int[] nums, int k) {
        return SubarraysWithMostKDistinct(nums, k) - SubarraysWithMostKDistinct(nums, k-1);
    }
    
    
    public int SubarraysWithMostKDistinct(int[] nums, int k)
    {
        var dict = new Dictionary<int, int>();
        var l = 0;
        var ans = 0;
 
        for (int r = 0; r < nums.Length; r++)
        {
            if (!dict.ContainsKey(nums[r]))
                dict.Add(nums[r], 0);
            dict[nums[r]]++;
 
            while (dict.Count > k)
            {
                if (dict.ContainsKey(nums[l]))
                {
                    dict[nums[l]]--;
                    if (dict[nums[l]] == 0)
                        dict.Remove(nums[l]);
                    l++;
                }
            }
            ans += (r - l + 1);
        }
 
        return ans;
    }

}

File: 0992_Subarrays with K Different Integers/SubarrayswithKDifferentIntegers_2.java
class Solution {
    public int subarraysWithKDistinct(int[] nums, int k) {
        return subarraysWithAtMostK(nums, k) - subarraysWithAtMostK(nums, k-1);
    }
    
    public int subarraysWithAtMostK(int[] nums, int k) {
        int ans = 0;
        Map<Integer,Integer> map = new HashMap<>();
        int l = 0;
        for(int r=0; r<nums.length; r++) {
            map.put(nums[r], map.getOrDefault(nums[r], 0) + 1);
            while(map.size() > k) {
                if(map.containsKey(nums[l])) {
                    int v = map.get(nums[l]);
                    if(v == 1) 
                        map.remove(nums[l]);
                    else 
                        map.put(nums[l], v-1);
                }
                l++;
            }
            
            ans += r - l + 1;
        }
        
        return ans;
    }
}